### ISSUE TITLE
str2var crashes the application when processing a non-convertible string - not very useful behavior

### ISSUE THREAD

**Issue by blurymind** [godotengine/godot#13021]
A common use case for str2var for me is to have it go through a collection of string variables and have it convert them to the appropriate type.
This however becomes a problem, as str2var will crash godot if it encounters a variable that is supposed to be a string in the first place.

I think it would be more useful if the function simply returns the original input string in that case, rather than crashing the engine. 

I made an ugly workaround function to get around that- just to stop it from crashing:

```gdscript
func cusStr2Var(inputString):
	if !isItAString(inputString):
		return str2var(inputString)
	else: return inputString

func isItAString(string): ##fixed
	var NameRegEx = RegEx.new()
	NameRegEx.compile('[A-Za-z]+')
	if NameRegEx.search(string) != null:
		return true
	else: return false
```
Its messy, might have to rewrite it, the idea is to avoid crashing

But would really like it if I didn't have to :) 
If you want to keep the crash message as to why it failed to convert it, why not turn it into a warning instead?

**Comment by blurymind** [comment]
One thing that could be nice is to at least have it return a null instead of blowing up. That way I can avoid using a regular expression to guess

**Comment by vnen** [comment]
For one side it's good to have it error, since it can avoid bugs. OTOH this current behavior makes it impossible to recovery, since Godot offers no error handling.

I don't like returning the original string, it makes impossible to know if it was a success or not. TBH this is also the case for `null`, it's totally possible to the original value was a `Null` since `typeof(str2var('null')) == TYPE_NIL` is `true`.

Also note that for a string to parsed correctly, it needs to be enclosed in double-quotes: `print(str2var('"This will print nicely"'))`.

**Comment by blurymind** [comment]
@vnen I agree that having some kind of error feedback is nice to have, but can't we make it a warning rather than a crash somehow? In many cases the extracted text doesn't have the the extra quotation marks and the current design breaks it. When you put it that way- yes the current design does make sense, but it is not convenient- especially when gdscript by design has avoided a 'try' command like the one that can be found in python.
So there is no way for me to catch the code exploding during the var2str operation - I am forced to mess with the data before it gets to it. If we cant return anything to indicate that it failed, can we at least have it not crash godot?

How about this, what if we pass an optional parameter to return something if it fails:
`var2str(inputVar,returnIfFailsVar)`
so example use can be

```
var myVar ="this might blow up"
var2str(myVar) #explodes as usual
var2str(myVar ,null) # returns null instead of exploding
var2str(myVar ,myVar )# returns original input instead of exploding
```

I can't think of anything else to suggest, I can close this issue if I am the only one who thinks it can be problematic sometimes. 

Perhaps someone will have a better idea to approach this?

**Comment by willnationsdev** [comment]
I'm not quite sure of the internals, but wouldn't passing no second parameter end up being equivalent to passing a null second parameter? The parameters don't end up in an array where you can distinctly count the size. They ultimately call a C++ function with actual Variants as parameters. You might have to completely rewrite `strtovar` to account for that, and even then, you'd have to put it under a different name to avoid naming conflicts since I don't think function overloading is an option here(?).

**Comment by blurymind** [comment]
@willnationsdev the general idea is to optionally stop godot from crashing when a non-convertible string has been passed to it

The second parameter could just as well be a simple boolean to tell it that some of the informationg that is being parsed to it might be intended to be a string to begin with.

So if it can't convert it to something, at least don't crash

**Comment by willnationsdev** [comment]
I'm not necessarily opposed to changing things to prevent a crash, but I don't think a second parameter is the answer in this case. A more efficient solution might be to have it return an ERR_* integer in case of failure. Doesn't seem as though it would be that difficult to implement.

**Comment by blurymind** [comment]
@willnationsdev I think we talked about that solution with @vnen and he brought up a very good point that str2var returns different type of variables and it would be a problem if it returns anything- I suggested that it returns a null. 
Your proposal is similar and has the same problem - if it returns an error integer, there is still a good chance that its the same value as an integer that you are converting from a string 

I think that it should return the original string - as it detected that it is a string that can not be converted and also give some sort of a warning in debugger that it failed to convert it - instead of crashing the entire application

**Comment by willnationsdev** [comment]
Hmm. I wonder if you could just test whether wrapping the content in double quotes creates a successful conversion, and then if so, return that. I haven't looked at the source code, but that would be an interesting approach. If you did it as a last resort, then you'd be able to just supply raw text to it and it would automatically output it as a string because it failed to match it against the string syntax and all of the other object syntaxes

**Comment by vnen** [comment]
Actually, returning the original string sounds like the best solution. You can compare the input and the output and if they're the same it means the parsing failed.

**Comment by willnationsdev** [comment]
If that's the case, the error might be fixed with modifications to `core/variant_parser.cpp`, line 1255.

    } else {
        r_err_str = "Expected value, got " + String(tk_name[token.type]) + ".";
        return ERR_PARSE_ERROR;
    }

...and subsequent changes to `_parse_array` and `_parse_dictionary`.

**Comment by vnen** [comment]
No, I think it should be changed only for GDScript (since this is a GDScript-only function), so just change this if block to just return the original argument instead of creating an error:

https://github.com/godotengine/godot/blob/1fa9aac3e415f53a095e955c8a37000629d56dde/modules/gdscript/gdscript_functions.cpp#L702-L708

People using the VariantParser directly can detect the error themselves.

**Comment by vnen** [comment]
The only problem with this change is that if people don't check the return value, they might let a parsing error slip by, so I'm still not sure if this is wanted. `str2var()` is intended to be used with the result from `var2str()`, it should never fail.

**Comment by blurymind** [comment]
@vnen  Godot's debugger is capable of printing warning messages - when for example a signal is fired to run a non-existent function - the application will continue running but a warning will be printed in the debugger. Why not use a warning instead of the termination?

Alternatively it could be optional to run str2var with a parameter to disable it from terminating the application and the current behavior be kept as the default.

It would be great to be able to just stream a bunch of strings to str2var and not be worried that one of them will crash the application

**Comment by Anutrix** [comment]
Was this fixed in #22934?

**Comment by akien-mga** [comment]
Should be yeah, closing as fixed by #22934. Please comment if you can still reproduce it.
